### ISSUE TITLE
POST request causes `input has been closed!` error with Sinatra

### ISSUE THREAD

**Issue by genki** [socketry/protocol-http#31]
I am using Falcon-0.42.3 with Sinatra-3.2.0.
No problems with GET requests but the POST causes the error.
The backtrace is here:
```
Stream is not readable, input has been closed!
vendor/ruby/3.2.0/gems/protocol-rack-0.2.4/lib/protocol/rack/input.rb:91:in `read_next'
vendor/ruby/3.2.0/gems/protocol-http-0.24.1/lib/protocol/http/body/stream.rb:69:in `read'
vendor/ruby/3.2.0/gems/rack-2.2.7/lib/rack/request.rb:447:in `POST'
vendor/ruby/3.2.0/gems/rack-2.2.7/lib/rack/request.rb:469:in `params'
vendor/ruby/3.2.0/gems/rack-2.2.7/lib/rack/request.rb:32:in `params'
vendor/ruby/3.2.0/gems/sinatra-3.0.6/lib/sinatra/base.rb:79:in `params'
vendor/ruby/3.2.0/gems/sinatra-3.0.6/lib/sinatra/base.rb:1126:in `dispatch!'
vendor/ruby/3.2.0/gems/sinatra-3.0.6/lib/sinatra/base.rb:949:in `block in call!'
vendor/ruby/3.2.0/gems/sinatra-3.0.6/lib/sinatra/base.rb:1109:in `catch'
vendor/ruby/3.2.0/gems/sinatra-3.0.6/lib/sinatra/base.rb:1109:in `invoke'
vendor/ruby/3.2.0/gems/sinatra-3.0.6/lib/sinatra/base.rb:949:in `call!'
vendor/ruby/3.2.0/gems/sinatra-3.0.6/lib/sinatra/base.rb:938:in `call'
```

**Comment by ioquatix** [comment]
Do you mind sharing what is doing the post request?

**Comment by genki** [comment]
@ioquatix 
Here is a tiny reproduction code.

```ruby
require "sinatra"

get "/" do
  <<~EOH
  <html>
  <body>
  <h1>Hello, world!</h1>
  <form action="/" method="post">
    <input type="submit" value="Submit">
  </form>
  EOH
end

post "/" do
  "Hello, world?"
end
```

And the Gemfile:
```ruby
source "https://rubygems.org"
gem "sinatra"
gem "falcon"
```

**Comment by genki** [comment]
And the screenshot
https://user-images.githubusercontent.com/4876/234249387-990154d3-7cda-4f19-869c-534ed56c6987.mov

**Comment by genki** [comment]
Had tested same reproduction code by `bundle exec ruby app.rb` on WEBrick, thin and puma with success.
My env is
 * M2 Macbook (MacOS 13.3)
 * ruby 3.2.1 (2023-02-08 revision 31819e82c8) [arm64-darwin22]

**Comment by ioquatix** [comment]
Thanks that is helpful, I will take a look.

**Comment by genki** [comment]
Found that the form with any input item causes no errors.
Treatment of the POST of a blank form may be the reason.

**Comment by ioquatix** [comment]
Sorry, I have not had a chance to carefully review your PRs yet. I will endeavour to do so within the next week or two. RubyKaigi preparation and work is very busy right now.

**Comment by genki** [comment]
@ioquatix don't worry ;)

**Comment by ioquatix** [comment]
You can help me if you have time by answering some questions.

My original assumption was that `@body = nil` means the body is not present OR closed. When the post request has no body, why should read return `""` instead of closed? i.e. should we consider these separate cases, semantically?

Recently I merged into Rack, `env['rack.input']` should be optional. In the case there is no body, it should not be presented. How does your proposal fit with this change?

If we decided to introduce `@closed`, we make the semantics more complex. Are there any other solutions that keep a more simple semantic?

I am okay to introduce such a change if it's required to correctly model the semantics, but I want to make sure we aren't just modelling accidental complexity.

**Comment by genki** [comment]
@ioquatix 
Sorry for delay, unfortunately I too have less time because of fixing things caused by Twitter API change.

As I mentioned above, the HTML form permit to send the POST request with no body.
This must not cause the IOError.
I think HTML is more basic than the Rack, so it is better to make fit Rack to HTML.

To avoid raising IOError when the body is empty, we have to distinguish the body is nil or empty.
Adding `@close` instance variable is just one of solutions, but something similar complexity would be needed, even if selecting others.

**Comment by ioquatix** [comment]
Fixed in v0.24.3.
